### PR TITLE
Initial Upload From IDE

### PR DESCRIPTION
**Pull request by me-no-dev**
For Test ONLY *** Requires Python to be installed and in the PATH!
After uploading the DNS_SD_Arduino_OTA example in the mDNS server examples through the regular method, your ESP should be visible in the ports selection (well give it a few)
When you select the device also make sure you select from the menu "Flash Using OTA".
change the sketch to your liking and give it another go, this time over the air :)

The python script and example are created by @igrr 
The uploader has some issues still, so expect it to not be 100% reliable.

### CODE REVIEW

File: hardware/esp8266com/esp8266/libraries/ESP8266mDNS/examples/DNS_SD_Arduino_OTA/DNS_SD_Arduino_OTA.ino [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    MDNS.begin(host);
+    MDNS.addService("arduino", "tcp", aport);
+    listener.begin(aport);
+    Serial.print("IP address: ");
+    Serial.println(WiFi.localIP());
+  }
+}
+
+void loop() {
+  if (ota_cmd_listener.parsePacket()) {
```
Comment (igrr) [inline hardware/esp8266com/esp8266/libraries/ESP8266mDNS/examples/DNS_SD_Arduino_OTA/DNS_SD_Arduino_OTA.ino]: ota_cmd_listener -> listener

Comment (me-no-dev) [inline hardware/esp8266com/esp8266/libraries/ESP8266mDNS/examples/DNS_SD_Arduino_OTA/DNS_SD_Arduino_OTA.ino]: fixed

**Comment by reaper7** [comment]
it works!!! :dancer: great idea

Esp.cpp -> updateSketch
seems to be:
```
    if (size > (getSketchSize()+getFreeSketchSpace())){
        if(restartOnFail) ESP.restart();
        return false;
    }
```
instead
```
    if (size > getFreeSketchSpace()){
        if(restartOnFail) ESP.restart();
        return false;
    }
```
with this changes works perfect for me :dancers:

**Comment by me-no-dev** [comment]
hmm... isn't this in the sketch? maybe older version?
also keep in mind that "done!" does not mean success yet :)
it means that it sent the code to the ESP, which can still fail...
working on it though

**Comment by reaper7** [comment]
Esp.cpp from this commit:
https://github.com/me-no-dev/Arduino/commit/0238e2c7aa1e87cdf6c424ce5e20cb0c3d19f6e3


success, success :)
I see new texts via Serial.println(....)

**Comment by sticilface** [comment]
I've given it a go, but it does not appear in the ports section... given it multiple goes, and waited.  I can upload to it fine using the python script so must be the DNS advertising... bit.  Any tips on debugging.. ?

Im using OS X.

**Comment by me-no-dev** [comment]
I am also on OS X and have 0 issues with mDNS and DNS-SD. I even see it in Safari when I advertise http service. Has to be something else you are missing.

**Comment by sticilface** [comment]
I've tried a different router to see if that might have been it.  nope..
I've tried two ESPs..

I added this line 
```MDNS.addService("arduino2", "http", 80);```  not sure if its right to test for bonjour in safari.. but that did not work either... 

Got any suggestions... ?  

bonjour is working as i've got other services.. my nas, printer etc working...

**Comment by me-no-dev** [comment]
for the esp to be seen by the IDE, the line is: MDNS.addService("arduino", "tcp", 8266);
for the esp to be seen by Safari, the line is: MDNS.addService("http", "tcp", 80);

**Comment by me-no-dev** [comment]
arduino/http is the name of the service (that name matters)
tcp is the network protocol
then the port that the service runs on

**Comment by me-no-dev** [comment]
and you need to restart safari after the ESP boots and sets up everything

**Comment by sticilface** [comment]
and that was the crucial bit of info.... restart safari... it works... and most importantly... restart the IDE... and boom!  

thank you.  and fantastic work.  I had a look at the mDNS library... and i've not got a clue what the hell is going on!

**Comment by sticilface** [comment]
although i get this board password thing...
https://www.dropbox.com/s/ho65x9olvnpx10d/Screenshot%202015-07-02%2022.05.05.png?dl=0

**Comment by me-no-dev** [comment]
it's OK, enter anything or nothing for now. We will implement password protected uploads and console, and that is the reason I left the dialog box.

**Comment by sticilface** [comment]
there must be a different issue then..  as i get failed with nothing / anything...
https://www.dropbox.com/s/ask0qirnzjvp3x4/Screenshot%202015-07-02%2022.54.52.png?dl=0

upload using the script works fine

**Comment by me-no-dev** [comment]
what sketch are you using? the OTA example?

**Comment by sticilface** [comment]
yes.  the OTA example.

**Comment by me-no-dev** [comment]
I just pushed a whole lot of changes. I advise you to get the latest example and changes and give it another go.
you only need to change ssid/pass in the example

**Comment by sticilface** [comment]
I saw that, changes had been made and I'm already in the process of cloning!

**Comment by sticilface** [comment]
It works.  good work!

**Comment by igrr** [comment]
Partially merged in #512.

**Comment by igrr** [comment]
Rebased the remaining changes related to WebServer request parsing directly on the esp8266 branch.
Closing this one, thanks.

**Pull request closed without merging**